### ISSUE TITLE
[java] New rule to enforce clarifying parentheses

### ISSUE THREAD

**Issue by Vampire** [pmd/pmd#6275]
**Proposed Rule Name:** `MissingClarifyingParentheses`

**Proposed Category:** Code Style

**Description:**

The `UselessParentheses` rule has an option to ignore clarifying parentheses which is on by default.
I'd like to even have the counterpart, a rule that enforces clarifying parentheses to be added to improve expression readability.

**Code Sample:**

```java
boolean foo, bar, baz, bam, boo, baa;
foo = bar = baz = bam = boo = new Random().nextBoolean();

baa = foo ? bar : baz ? bam : boo; // bad 1
baa = foo ? bar ? baz : bam : boo; // bad 2

baa = foo ? bar : (baz ? bam : boo); // fix for bad 1
baa = foo ? (bar ? baz : bam) : boo; // fix for bad 2
```

**Possible Properties:**

*   according to `ignoreBalancing` on `UselessParentheses`, balancing parens could also be considered clarifying here and be enforced, but configurable whether they should be considered or not.

**Comment by UncleOwen** [comment]
Is there a limit to how many clarifying parentheses you want? For example, would you want this rule to complain about
```java
IO.println(3 + 4 * 5);
```
?

Is there a reason you don't want your fixed lines to be
```java
baa = (foo ? bar : (baz ? bam : boo)); // fix for bad 1
baa = (foo ? (bar ? baz : bam) : boo); // fix for bad 2
```
?

**Comment by UncleOwen** [comment]
Answering my own question: We could use the same definition of "clarifying" as the existing `UselessParentheses` rule does. Sorry for the noise, that one was too obvious.

**Comment by Vampire** [comment]
> Is there a limit to how many clarifying parentheses you want? 

No, I always want the clarifying parens.

> For example, would you want this rule to complain about

Yes

> Is there a reason you don't want your fixed lines to be

Yes, those outer ones are useless and do not clarify anything.
Actually, I'd expect the `UselessParentheses` rule to complain about the outer pair.
